Relative value bonus for new rides.

GrahamRCT opened the pull request:
This prevents, for instance, charging $6 for any new ride, regarldess of its quality.

duncanspumpkin commented on src/openrct2/ride/ride_ratings.c:
```diff
@@ -618,15 +618,27 @@ static void ride_ratings_calculate_value(rct_ride *ride)
+	if (monthsOld <= 4)
+		value = (value * 3) / 2;   // 1.5x
+
+	else if (monthsOld <= 12) 
```
I don't like this sort of formatting. Use brackets for this as the else does not look like its part of the previous if. 

I think we could really do with refactoring this whole function into a table with upper lower month bands and multipliers (+bonuses for original ratings) for these bands. This would mean it would be much easier to see how your version and the previous worked. 

Existing:
age max | multiplier | divisor | bonus
----------|-----------|-------|---------
4 | 1 | 1 | 30
12 | 1 | 1 | 10
63 | 3 | 4 | 0
87 | 1 | 2 | 0
103 | 1 | 4 | 0
119 | 1 | 16 | 0
127 | 1 | 32 | 0
max | 1 | 64 | 0

GrahamRCT commented:
I made a table-based approach using very basic 2-dimensional arrays. Computation-wise it should be very fast. Readability-wise I'm not sure whether it's up to standards.

duncanspumpkin commented:
It doesn't need to be fast this isn't recalculated every tick. I would go for readability over speed. I would suggest creating a new struct to increase readability further.

GrahamRCT commented:
I didn't (yet) add structs. For one I didn't feel they represent a table very well.

That said, this change should make it much more readable.

Performance wise, the performance grows linearly with the amount of different age modifiers.

duncanspumpkin commented on src/openrct2/ride/ride_ratings.c:
```diff
@@ -610,32 +610,71 @@ static void ride_ratings_calculate(rct_ride *ride)
 static void ride_ratings_calculate_value(rct_ride *ride)
 {
+    static const int age_table_new[][4] = {
+        //Monhts    Multiplier  Divisor     Summand
```
`Months`

I think it might simplify the logic if you did it the other way round start at the minimum age.

duncanspumpkin commented on src/openrct2/ride/ride_ratings.c:
```diff
@@ -610,32 +610,71 @@ static void ride_ratings_calculate(rct_ride *ride)
 		return;
 	}
 
+	// Start with the base ratings, multiplied by the ride type specific weights for excitement, intensitve and nausea.
```
`intensity`

duncanspumpkin commented on src/openrct2/ride/ride_ratings.c:
```diff
@@ -610,32 +610,71 @@ static void ride_ratings_calculate(rct_ride *ride)
-			value += 20;
-	}
+	const int (*age_tab)[4] = age_table_new;
+	int age_mult_size = sizeof(age_table_new)/sizeof(int)/4;
```
We have a `count_of` function that will do the size of an array.

janisozaur replied:
[`countof`](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/common.h#L112)

duncanspumpkin commented on src/openrct2/ride/ride_ratings.c:
```diff
@@ -610,32 +610,71 @@ static void ride_ratings_calculate(rct_ride *ride)
 
 static void ride_ratings_calculate_value(rct_ride *ride)
 {
+    static const int age_table_new[][4] = {
```
A struct would mean you wouldn't need to add the comments and make it more legable.
```
struct{
    int months;
    int multiplier;
    int divisor;
    int summand;
};
```

GrahamRCT commented:
What about this?

I dunno if there's a better way in C  to initialize an array of structs that also looks like a table.

duncanspumpkin commented:
Looks fine. I might have chosen a different name for the struct but its not a big issue.

janisozaur commented on src/openrct2/ride/ride_ratings.c:
```diff
@@ -610,32 +610,76 @@ static void ride_ratings_calculate(rct_ride *ride)
+		value = (value * last_row.multiplier) / last_row.divisor + last_row.summand;
+	}
+	else {
+		int it;
```
move the variable declaration inside the loop

GrahamRCT replied:
So C99 is available?

janisozaur replied:
[sure](https://github.com/OpenRCT2/OpenRCT2/blob/1876785/CMakeLists.txt#L292)

janisozaur commented on src/openrct2/ride/ride_ratings.c:
```diff
@@ -610,32 +610,76 @@ static void ride_ratings_calculate(rct_ride *ride)
-		if (monthsOld <= 4)
-			value += 20;
-	}
+	const row (*age_table) = age_table_new;
```
why is there a parentheses here?

GrahamRCT replied:
That's a leftover from when this was a pointer to an array. I'll remove those.

janisozaur commented on src/openrct2/ride/ride_ratings.c:
```diff
@@ -610,32 +610,76 @@ static void ride_ratings_calculate(rct_ride *ride)
+		{104,       27,         64,         0},   // 0.42x
+		{120,       81,         256,        0},   // 0.32x
+		{128,       81,         512,        0},   // 0.16x
+		{200,       81,         1024,       0},   // 0.08x
```
How does this work? The way I see the code below it will always select the `0.56x "easter egg"` value and never this one. Can you provide conditions under which this entry will be selected?

duncanspumpkin replied:
If the ride is super old i.e. >200 months it enters easter egg mode. Otherwise it is within the age range with the months representing what it has to be younger than.

janisozaur replied:
right, that checks.

GrahamRCT replied:
The last row is evaluated differently. 

```
 	// Ride is older than oldest age in the table?
 	if(monthsOld >= last_row.months) {
 		value = (value * last_row.multiplier) / last_row.divisor + last_row.summand;
 	}because it represents the 
```

GrahamRCT commented:
By the way, should I use sint32 instead of int?

IntelOrca commented:
@GrahamRCT yes, it ensures the logic is consistent across different platforms.

GrahamRCT commented:
> yes, it ensures the logic is consistent across different platforms.

OK, updated.


Also, I added some test print-outs and compared the values of all rides in the RCT2 title sequence. If anyone's interested, I've attached the reports.

[values.zip](https://github.com/OpenRCT2/OpenRCT2/files/867614/values.zip)

GrahamRCT commented:
The _discussion_ label was added a week ago, so I assume there's still doubt as to whether this change would be a positive thing.

Any specific thoughts or objections I could address?

Perhaps a simpler before/after comparison helps:

**Before**
Brand new rides (<4 months) get a +$6 bonus
New rides (<12 months) get a +2$ bonus

**After**
Brand new rides (<4 months) get a 1.5× bonus
New rides (<12 months) get a 1.2× bonus


So, for a brand new ride that guests would normally pay $2 for:
Before: $8
After: $3

A brand new ride that guests would normally pay $10 for:
Before: $16
After: $15

duncanspumpkin commented:
Yeah I put the discussion label on as there were some questions asked on gitter about if we want to make these changes. Thanks for writing out an explanation. Personally I think this change is appropriate and reasonable and makes sense. The main negative is that it will affect the balance of the game but in some ways i think this will make the game more deterministic and better.
